hca/define_api.py

```diff
@@ -1,37 +1,172 @@
+    def __init__(self, test=False):
         """Initialize the CLI API."""
-        self.spec = self.get_spec()
-        self.base_url = self.spec['host'] + self.spec['basePath']
-        self.parser = get_parser(self.spec)
+        spec = self.get_spec(test)
+        scheme = spec["schemes"][0] if "schemes" in spec else "https"
+        self.base_url = "{}://{}{}".format(scheme, spec['host'], spec['basePath'])
+        self.parser, self.param_data = get_parser(spec)
 
-    def get_spec(self):
-        """Load the API specification."""
-        with open(os.path.dirname(os.path.realpath(__file__)) + "/api_spec.json") as fp:
+    def get_spec(self, test=False):
+        """
+        Load the API specification.
+
+        :param test: boolean flag to indicate which spec to use.
+                     api_spec is the spec downloaded from the api on build.
+                     ../test/test is the mocked up spec I've been toying with.
+        :return:     The dictionary containing all swagger specification definitions.
+        """
+        url = os.path.dirname(os.path.realpath(__file__)) + "/api_spec.json"
+        if test:
+            url = os.path.dirname(os.path.realpath(__file__)) + "/../test/test.json"
+
+        with open(url) as fp:
             api_spec_dict = json.load(fp)
         return api_spec_dict
 
+    def _build_url(self, endpoint, namespace):
+        """Add positional arguments into the url."""
+        split_endpoint = endpoint.split("-")[1:]
+
+        all_positional_args_for_endpoint = [arg['argument'] for arg in self.param_data[endpoint]['positional']]
+
+        # If the api needs file/write/{sdf} functionality, will become put-file-write
+        given_positional_args = [p for p in split_endpoint]
+        for positional_arg in all_positional_args_for_endpoint:
+            if positional_arg in namespace:
+                arg = namespace[positional_arg]
+                given_positional_args.append(arg)
+        url = self.base_url + "/" + "/".join(given_positional_args)
+        return url
+
+    def _add_arg(self, arg, payload, hierarchy):
+        """
+        Add a given argument to its associated payload.
+
+        Arguments:
+         - arg: The value that is being assigned.
+         - payload: The payload structure before adding this value.
+         - hierarchy: Data structure that describes where in payload to put arg.
+        """
+        inner_payload = payload
+
+        # Iterating through payload structure to find the right value to set.
+        while len(hierarchy) > 1:
+            curr_level = hierarchy[0]
+
+            if curr_level not in inner_payload:
+                inner_payload[curr_level] = [] if isinstance(hierarchy[1], list) else {}
+            inner_payload = inner_payload[curr_level]
+
+            hierarchy = hierarchy[1:]
+
+        curr_level = hierarchy[0]
+
+        # Command line argument represents a list of objects.
+        if type(curr_level) == list and len(curr_level) > 0:
```

Comment:
typically `instanceof` is a safer thing to do, because `type(..)` requires an exact match of types.